scio-core/src/main/scala/com/spotify/scio/hash/MutableScalableBloomFilter.scala

```diff
@@ -0,0 +1,215 @@
+    val dis = new DataInputStream(bais)
+
+    val fpProb = dis.readDouble()
+    val headCapacity = dis.readLong()
+    val growthRate = dis.readInt()
+    val tighteningRatio = dis.readDouble()
+    val headFPProb = dis.readDouble()
+    val headCount = dis.readLong()
+    val numFilters = dis.readInt()
+    val head = if (numFilters > 0) Some(g.BloomFilter.readFrom[T](dis, funnel)) else None
+    val tail: List[Either[g.BloomFilter[T], SerializedBloomFilters]] = {
+      if (numFilters > 1) {
+        val baos = new ByteArrayOutputStream()
+        ByteStreams.copy(dis, baos)
+        List(Right(SerializedBloomFilters(numFilters - 1, baos.toByteArray)))
+      } else {
+        List.empty
+      }
+    }
+    MutableScalableBloomFilter[T](
+      fpProb,
+      headCapacity,
+      growthRate,
+      tighteningRatio,
+      headFPProb,
+      headCount,
+      head,
+      tail
+    )
+  }
+
+  implicit def coder[T](implicit funnel: g.Funnel[T]): Coder[MutableScalableBloomFilter[T]] =
+    Coder.xmap[Array[Byte], MutableScalableBloomFilter[T]](Coder.arrayByteCoder)(
+      bytes => fromBytes[T](bytes)(funnel),
+      sbf => toBytes[T](sbf)
+    )
+}
+
+case class SerializedBloomFilters(numFilters: Int, filterBytes: Array[Byte]) {
+  def deserialize[T](implicit funnel: g.Funnel[T]): List[g.BloomFilter[T]] = {
+    val bais = new ByteArrayInputStream(filterBytes)
+    (1 to numFilters).map(_ => g.BloomFilter.readFrom[T](bais, funnel)).toList
+  }
+}
+
+/**
+ * @param fpProb            The desired false positive probability
+ * @param headCapacity      The capacity of the filter at the head of `filters`
+ * @param growthRate        The growth rate of each subsequent filter added to `filters`
+ * @param tighteningRatio   The tightening ratio applied to the current `fpProb` to maintain the false positive probability over the sequence of filters
+ * @param head              The underlying bloom filter currently being inserted into, or `None` if this scalable filter has just been initialized
+ * @param tail              The underlying already-saturated bloom filters, lazily deserialized from bytes as necessary.
+ * @param headFPProb        The false positive probability of the head of `filters`
+ * @param headCount         The number of items currently in the filter at the head of `filters`
+ * @param funnel            The funnel to turn `T`s into bytes
+ * @tparam T                The type of objects inserted into the filter
+ */
+case class MutableScalableBloomFilter[T](
+  fpProb: Double,
+  private var headCapacity: Long,
+  private val growthRate: Int,
+  private val tighteningRatio: Double,
+  private var headFPProb: Double,
+  // storing a count of items in the head avoids calling the relatively expensive `approximateElementCount` after each insert
+  private var headCount: Long,
+  private var head: Option[g.BloomFilter[T]],
+  private var tail: List[Either[g.BloomFilter[T], SerializedBloomFilters]]
```

Comment:
If it's not serialized yet, `tail` should be `List(Right(SerializedBloomFilters(...)))` i.e. always `List` of a single element? Can this type be `Either[List[g.BF[T]], SerializedBF]]` so it's clearer?

Reply:
Let's say you've pulled an existing SBF off disk and you're adding new items, then you saturate the `head` and need to add another BF, pushing the current head to `tail`.At that point, if we have `Either[List[g.BF[T]], SerializedBF]]` we'll be forced to serialize the old `head` which seemed like a costly operation to have on insert, so here it's deferred until you actually want to encode to bytes.